hw/bsp/nutiny_sdk_nuc505/nutiny_sdk_nuc505.c

```diff
@@ -0,0 +1,118 @@
+/* 
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Ha Thach (tinyusb.org)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * This file is part of the TinyUSB stack.
+ */
+
+#include "bsp/board.h"
+#include "NUC505Series.h"
+
+#define LED_PORT     PC
+#define LED_PIN      3
+#define LED_STATE_ON 0
+
+void board_init(void)
+{
+  /* Enable XTAL */
+  CLK->PWRCTL |= CLK_PWRCTL_HXTEN_Msk;
+
+  CLK_SetCoreClock(96000000);
+
+  /* Set PCLK divider */
+  CLK_SetModuleClock(PCLK_MODULE, 0, 1);
+
+  /* Update System Core Clock */
+  SystemCoreClockUpdate();
+
+  /* Enable USB IP clock */
+  CLK_EnableModuleClock(USBD_MODULE);
+
+  /* Select USB IP clock source */
+  CLK_SetModuleClock(USBD_MODULE, CLK_USBD_SRC_EXT, 0);
+
+  CLK_SetModuleClock(PCLK_MODULE, 0, 1);
+
+  /* Enable PHY */
+  USBD_ENABLE_PHY();
+  /* wait PHY clock ready */
+  while (1) {
+      USBD->EP[EPA].EPMPS = 0x20;
+      if (USBD->EP[EPA].EPMPS == 0x20)
+          break;
+  }
+
+  /* Force SE0, and then clear it to connect*/
+  USBD_SET_SE0();
+
+#if CFG_TUSB_OS  == OPT_OS_NONE
+  // 1ms tick timer
+  SysTick_Config(96000000 / 1000);
+#endif
+
+  GPIO_SetMode(LED_PORT, 1UL << LED_PIN, GPIO_MODE_OUTPUT);
+}
+
+#if CFG_TUSB_OS  == OPT_OS_NONE
+volatile uint32_t system_ticks = 0;
+void SysTick_Handler (void)
+{
+  system_ticks++;
+}
+
+uint32_t board_millis(void)
+{
+  return system_ticks;
+}
+#endif
+
+//--------------------------------------------------------------------+
+// Board porting API
+//--------------------------------------------------------------------+
+
+void board_led_write(bool state)
+{
+  uint32_t current = (state) ? LED_STATE_ON : (1-LED_STATE_ON);
+  current <<= LED_PIN;
+  uint32_t irq_state = __get_PRIMASK();
+  __disable_irq();
```

Comment:
May I ask why we need to disable irq and to write to GPIO. Sorry if the question is so obvious with this mcu

Reply:
The LED is being changed via a read-modify-write cycle. This would cause a concurrency issue if the task were switched out to another task (or interrupt) that also writes to this IO port. It looks like the NUC505's GPIO doesn't have a 'bit set' or 'bit reset' register, so this is required.

Reply:
@pigrew wow, that is new to me, thank you very much for the detailed explanation. NUC505 is definitely a bit more difficult to work with :D. I could only imagine to use this mcu bit banging an protocol like i2c :face_with_head_bandage: